# Data Exploration

In this tutorial we will explore data in the MIMIC Waveform Database.

The **objectives** are:
- To gain an understanding of the structure of the MIMIC Database, consisting of subjects, studies, records, and segments.
- To start using the WaveForm DataBase Toolbox to work with waveform data files
- To understand how to find out what signals are present in a record or a segment, and how long they last.
- To gain experience in searching for records which contain the signals required for a particular study.

---
## Setup

### Specify the required Python packages
We'll import the following:
- _sys_: an essential python package
- _wfdb_: The WaveForm DataBase Toolbox package, which contains tools for processing waveform data such as those in MIMIC.
- _pathlib_ (well a particular function from _pathlib_, called _Path_)

In [21]:
import sys
import wfdb
from pathlib import Path

Now that we have imported these packages (_i.e._ toolboxes) we have a set of tools (functions) ready to use.

<div class="alert alert-block alert-warning"> <b>Resource:</b> You can find out more about the WFDB package <a href="https://physionet.org/content/wfdb-python/3.4.1/">here</a>. </div>

### Specify the version of the MIMIC Waveform Database to use
We will use either the:
- [MIMIC III Waveform Database](https://doi.org/10.13026/c2607m), or
- MIMIC IV Waveform Database

- Specify which version of the database to use:

In [22]:
mimic_wfdb_version = 4 # either: 3 or 4

- Setup for this particular version of the database:

In [23]:
if mimic_wfdb_version == 3:
    wfdb.set_db_index_url('https://physionet.org/files') # use the version of the WFDB toolbox which corresponds to MIMIC IV.
    database_name = 'mimic3wdb/1.0' # The name of the MIMIC III Waveform Database on Physionet (see URL: https://physionet.org/content/mimic3wdb/1.0/)
elif mimic_wfdb_version == 4:
    wfdb.set_db_index_url('https://challenge.physionet.org/benjamin/db') # use the version of the WFDB toolbox which corresponds to MIMIC IV.
    database_name = 'mimic4wdb/0.1.0' # The name of the MIMIC IV Waveform Demo Database on Physionet
print("Using MIMIC Waveform Database version: {}".format(mimic_wfdb_version))

Using MIMIC Waveform Database version: 4


---
## Identify the records in the database

### Get a list of records

- Use the [`get_record_list`](https://wfdb.readthedocs.io/en/latest/io.html#wfdb.io.get_record_list) function from the WFDB toolbox to get a list of records in the database.

In [24]:
records = []
no_records_to_load = 205
subjects = wfdb.get_record_list(database_name)
print("The '{}' database contains data from {} subjects".format(database_name, len(subjects)))
finished = False
for subject in subjects:
    if (not finished):
        studies = wfdb.get_record_list(f'{database_name}/{subject}')
        for study in studies:
            if mimic_wfdb_version == 3 & ("_" not in study): # This skips any files which don't contain waveform data (e.g. numerics files)
                continue
            records.append(Path(f'{subject}{study}'))
            if len(records) >= no_records_to_load: # stop if we've loaded enough records
                finished = True
if not finished:
    print("Done: Went through entire database and loaded list of {} records for '{}' database".format(len(records), database_name))
else:
    print("Done: Went through first bit of database and loaded list of {} records for '{}' database".format(len(records), database_name))

The 'mimic4wdb/0.1.0' database contains data from 198 subjects
Done: Went through entire database and loaded list of 200 records for 'mimic4wdb/0.1.0' database


### Look at the records

- Display the first few records

In [25]:
print("First five records: {}".format(records[0:5]))
if mimic_wfdb_version == 3:
    print("\nNote the formatting of these records:\n - intermediate directory ('30' in this case)\n - subject identifier ('3000003' in this case)\n - record identifier (e.g. '3000003_0001'")
else:
    print("\nNote the formatting of these records:\n - intermediate directory ('p100' in this case)\n - subject identifier (e.g. 'p10014354')\n - record identifier (e.g. '81739927'")

First five records: [PosixPath('p100/p10014354/81739927/81739927'), PosixPath('p100/p10019003/87033314/87033314'), PosixPath('p100/p10020306/83404654/83404654'), PosixPath('p100/p10039708/83411188/83411188'), PosixPath('p100/p10039708/85583557/85583557')]

Note the formatting of these records:
 - intermediate directory ('p100' in this case)
 - subject identifier (e.g. 'p10014354')
 - record identifier (e.g. '81739927'


<div class="alert alert-block alert-info"> <b>Q:</b> Can you print the names of the last five records? <br> <b>Hint:</b> in Python, the last five elements can be specified using '[-5:]' </div>

---
## Extract metadata for a record

Each record contains metadata stored in a header file, named "\<record name\>.hea"

### Specify the online directory containing a record's data

In [26]:
if mimic_wfdb_version == 3:
    no = 0
else:
    no = 3 # specify the fourth record (noting that in Python the first index is 0)
record = records[no]
record_dir = f'{database_name}/{record.parent}'
print("Physionet directory specified for this record: {}".format(record_dir))

Physionet directory specified for this record: mimic4wdb/0.1.0/p100/p10039708/83411188


### Specify the subject identifier

Extract the record name (e.g. '83411188') from the record (e.g. 'p100/p10039708/83411188/83411188'):

In [27]:
record_name = record.name
print("Record name: {}".format(record_name))

Record name: 83411188


### Load the metadata for this record
- Use the [`rdheader`](https://wfdb.readthedocs.io/en/latest/io.html#wfdb.io.rdheader) function from the WFDB toolbox to load metadata from the record header file

In [28]:
record_data = wfdb.rdheader(record_name, pn_dir=record_dir, rd_segments=True)
print("Done: metadata loaded for record '{}' from header file at URL: {}".format(record_name, "https://physionet.org/content/" + record_dir + "/" + record_name + ".hea"))
# NOTE "https://physionet.org/content/" won't be correct until the MIMIC-IV-Waveform is officially released

Done: metadata loaded for record '83411188' from header file at URL: https://physionet.org/content/mimic4wdb/0.1.0/p100/p10039708/83411188/83411188.hea


---
## Inspect details of physiological signals recorded in this record
- Printing a few details of the signals from the extracted metadata

In [29]:
print("- Number of signals: {}".format(record_data.n_sig))
print("- Duration: {:.1f} hours".format(record_data.sig_len/(record_data.fs*60*60))) 
# NOTE fs isn't 125 Hz
print("- Sampling frequency: {} Hz".format(record_data.fs))

- Number of signals: 6
- Duration: 14.2 hours
- Sampling frequency: 62.4725 Hz


---
## Inspect the segments making up a record
Each record is typically made up of several segments

### Inspect the files in a record

In [30]:
if mimic_wfdb_version == 3:
    files = wfdb.get_record_list(str(Path(record_dir)))
    print("Done: Loaded list of {} records for subject '{}'".format(len(files), str(Path(record_dir).parts[-1])))
else:
    files = wfdb.get_record_list(str(Path(record_dir).parent))
    print("Done: Loaded list of {} records for subject '{}'".format(len(files), str(Path(record_dir).parent.parts[-1])))

Done: Loaded list of 2 records for subject 'p10039708'


### Inspect the contents of the first two files

In [31]:
if mimic_wfdb_version == 3:
    print("The first file is: '{}', which provides metadata for the whole record".format(files[0]) )
    print("The second file is: '{}', which contains numerics data".format(files[1]) )    
else:
    print("The first file contains the first record: '{}'".format(files[0]) )
    print("The second file contains the second record: '{}'".format(files[1]) )
    print("where the number before the / is the record number and the number after is for the multi-segment header file. This multi-segment header file provides general information about the record.")

The first file contains the first record: '83411188/83411188'
The second file contains the second record: '85583557/85583557'
where the number before the / is the record number and the number after is for the multi-segment header file. This multi-segment header file provides general information about the record.


### Inspect the segments within a record

In [32]:
segments = record_data.seg_name
# --- these lines remove non-waveform data from MIMIC III files
for curr_seg in segments:
    if ("~" in curr_seg) or ("layout" in curr_seg):
        segments.remove(curr_seg)
# ---
print("The {} segments from record {} are: {}".format(len(segments), record_name, segments) )

The 6 segments from record 83411188 are: ['83411188_0000', '83411188_0001', '83411188_0002', '83411188_0003', '83411188_0004', '83411188_0005']


Note the format of the names of the files containing waveform data for each segment: record directory, "_", segment number

---
## Inspect an individual segment
### Read the metadata for this segment
- Read the metadata from the header file

In [33]:
segment_metadata = wfdb.rdheader(record_name=segments[2], pn_dir=record_dir) 
print("Header metadata loaded for the segment '{}' in study '{}' for subject '{}'".format(segments[1], record_name, str(Path(record_dir).parent.parts[-1])))

Header metadata loaded for the segment '83411188_0001' in study '83411188' for subject 'p10039708'


### Find out what signals are present

In [34]:
print("This segment contains the following signals: {}".format(segment_metadata.sig_name))
print("The signals are measured in units of: {}".format(segment_metadata.units))

This segment contains the following signals: ['II', 'V', 'aVR', 'ABP', 'Pleth', 'Resp']
The signals are measured in units of: ['mV', 'mV', 'mV', 'mmHg', 'NU', 'Ohm']


See [here](https://archive.physionet.org/mimic2/mimic2_waveform_overview.shtml#signals-125-samplessecond) for definitions of signal abbreviations.

<div class="alert alert-block alert-info"> <b>Q:</b> Which of these signals is still present in segment '3000003_0014' (for MIMIC III) or segment '83411188_0000' (for MIMIC IV)? </div>

### Find out how long each signal lasts

All signals in a segment are time-aligned, measured at the same sampling frequency, and last the same duration:

In [35]:
print("All the signals are sampled at {} Hz".format(segment_metadata.fs))
print("and they last for {:.1f} minutes".format(segment_metadata.sig_len/(segment_metadata.fs*60)) )

All the signals are sampled at 62.4725 Hz
and they last for 0.9 minutes


## Identify records suitable for analysis

The signals available and their durations vary from one record (and segment) to the next. Since most studies require specific types of signals (e.g. blood pressure and photoplethysmography signals), it is important to identify which records (or segments) contain the required signals for the required duration.

### Setup

In [36]:
import pandas as pd
from pprint import pprint

In [37]:
print("Earlier, we loaded a list of {} records for '{}' database".format(len(records), database_name))

Earlier, we loaded a list of 200 records for 'mimic4wdb/0.1.0' database


### Specify requirements

- Required signals

In [38]:
required_sigs = ['ABP', 'Pleth']

- Required duration

In [39]:
req_seg_duration = 10*60  # converting from minutes to seconds

### Find out how many records meet the requirements

In [49]:
matching_recs = {'dir':[],'seg_name':[],'length':[]}
for record in records:
    print('Record: {}'.format(record), end="", flush=True)
    record_dir = f'{database_name}/{record.parent}'
    record_name = record.name
    print(' (reading data)')
    record_data = wfdb.rdheader(record_name, pn_dir=record_dir, rd_segments=True)
    # Check whether all the required signals are present in the record
    sigs_present = record_data.sig_name
    if not all(x in sigs_present for x in required_sigs):
        print('   (missing signals)')
        continue
    # Get the segments for the record
    segments = record_data.seg_name
    # First check to see if the segment is 10 min long, if not move onto the next one
    gen = (segment for segment in segments if segment != '~')
    for segment in gen:
        print(' - Segment: {}'.format(segment), end="", flush=True)
        segment_metadata = wfdb.rdheader(record_name=segment, pn_dir=record_dir)
        seg_length = segment_metadata.sig_len/(segment_metadata.fs)
        if seg_length < req_seg_duration:
            print(' (too short at {:.1f} mins)'.format(seg_length/60))
            continue
        # Next check that all required signals are present in the segment
        sigs_present = segment_metadata.sig_name
        
        if all(x in sigs_present for x in required_sigs):
            matching_recs['dir'].append(record_dir)
            matching_recs['seg_name'].append(segment)
            matching_recs['length'].append(seg_length)
            print(' (met requirements)')
            # Since we only need one segment per record break out of loop
            break
        else:
            print(' (long enough, but missing signal(s))')

print("A total of {} records met the requirements:".format(len(matching_recs['dir'])))
#df_matching_recs = pd.DataFrame(data=matching_recs)
#df_matching_recs.to_csv('matching_records.csv', index=False)
#p=1

Record: p100/p10014354/81739927/81739927 (reading data)
   (missing signals)
Record: p100/p10019003/87033314/87033314 (reading data)
   (missing signals)
Record: p100/p10020306/83404654/83404654 (reading data)
 - Segment: 83404654_0000 (too short at 0.0 mins)
 - Segment: 83404654_0001 (long enough, but missing signal(s))
 - Segment: 83404654_0002 (too short at 0.1 mins)
 - Segment: 83404654_0003 (too short at 0.3 mins)
 - Segment: 83404654_0004 (long enough, but missing signal(s))
 - Segment: 83404654_0005 (met requirements)
Record: p100/p10039708/83411188/83411188 (reading data)
 - Segment: 83411188_0000 (too short at 0.0 mins)
 - Segment: 83411188_0001 (too short at 0.1 mins)
 - Segment: 83411188_0002 (too short at 0.9 mins)
 - Segment: 83411188_0003 (too short at 0.3 mins)
 - Segment: 83411188_0004 (too short at 0.3 mins)
 - Segment: 83411188_0005 (long enough, but missing signal(s))
Record: p100/p10039708/85583557/85583557 (reading data)
   (missing signals)
Record: p100/p10079700/

Record: p124/p12449182/84179115/84179115 (reading data)
   (missing signals)
Record: p124/p12470697/84213440/84213440 (reading data)
   (missing signals)
Record: p125/p12545470/80674263/80674263 (reading data)
   (missing signals)
Record: p125/p12563752/85393521/85393521 (reading data)
   (missing signals)
Record: p125/p12567919/82552643/82552643 (reading data)
   (missing signals)
Record: p127/p12789108/85835247/85835247 (reading data)
   (missing signals)
Record: p128/p12801061/80154334/80154334 (reading data)
   (missing signals)
Record: p128/p12807320/81739815/81739815 (reading data)
   (missing signals)
Record: p128/p12847457/83054955/83054955 (reading data)
   (missing signals)
Record: p128/p12872596/85230771/85230771 (reading data)
 - Segment: 85230771_0000 (too short at 0.0 mins)
 - Segment: 85230771_0001 (too short at 0.1 mins)
 - Segment: 85230771_0002 (too short at 0.3 mins)
 - Segment: 85230771_0003 (too short at 0.1 mins)
 - Segment: 85230771_0004 (met requirements)
Record

 - Segment: 89225487_0000 (too short at 0.0 mins)
 - Segment: 89225487_0001 (too short at 0.1 mins)
 - Segment: 89225487_0002 (too short at 7.4 mins)
 - Segment: 89225487_0003 (too short at 0.1 mins)
 - Segment: 89225487_0004 (too short at 0.1 mins)
 - Segment: 89225487_0005 (too short at 0.4 mins)
 - Segment: 89225487_0006 (too short at 0.1 mins)
 - Segment: 89225487_0007 (met requirements)
Record: p168/p16820602/82493254/82493254 (reading data)
   (missing signals)
Record: p168/p16846280/83877081/83877081 (reading data)
   (missing signals)
Record: p168/p16848565/84917468/84917468 (reading data)
   (missing signals)
Record: p169/p16949991/88967754/88967754 (reading data)
   (missing signals)
Record: p169/p16955095/82284982/82284982 (reading data)
   (missing signals)
Record: p169/p16971339/87311151/87311151 (reading data)
   (missing signals)
Record: p170/p17032851/84391267/84391267 (reading data)
 - Segment: 84391267_0000 (too short at 0.0 mins)
 - Segment: 84391267_0001 (met requir

 - Segment: 81937383_0049 (too short at 0.2 mins)
 - Segment: 81937383_0050 (long enough, but missing signal(s))
 - Segment: 81937383_0051 (too short at 0.1 mins)
 - Segment: 81937383_0052 (long enough, but missing signal(s))
 - Segment: 81937383_0053 (too short at 4.4 mins)
Record: p187/p18709273/88457020/88457020 (reading data)
   (missing signals)
Record: p187/p18742074/84505262/84505262 (reading data)
 - Segment: 84505262_0000 (too short at 0.0 mins)
 - Segment: 84505262_0001 (too short at 0.1 mins)
 - Segment: 84505262_0002 (too short at 0.1 mins)
 - Segment: 84505262_0003 (too short at 0.1 mins)
 - Segment: 84505262_0004 (too short at 0.0 mins)
 - Segment: 84505262_0005 (too short at 0.1 mins)
 - Segment: 84505262_0006 (too short at 0.1 mins)
 - Segment: 84505262_0007 (too short at 0.0 mins)
 - Segment: 84505262_0008 (too short at 0.1 mins)
 - Segment: 84505262_0009 (too short at 1.4 mins)
 - Segment: 84505262_0010 (met requirements)
Record: p187/p18768904/82650643/82650643 (read

In [56]:
print("A total of {} out of {} records met the requirements.".format(len(matching_recs['dir']), len(records)))
print("\nThe relevant segment names are:")
print(matching_recs['seg_name'])
print("\nThe corresponding directories are:")
print(matching_recs['dir'])


A total of 52 out of 200 records met the requirements.

The relevant segment names are:
['83404654_0005', '82924339_0007', '84248019_0005', '82439920_0004', '82800131_0002', '84304393_0001', '89464742_0001', '88958796_0004', '88995377_0001', '85230771_0004', '86643930_0004', '81250824_0005', '87706224_0003', '83058614_0005', '82803505_0017', '88574629_0001', '87867111_0012', '84560969_0001', '87562386_0001', '88685937_0001', '86120311_0001', '89866183_0014', '89068160_0002', '86380383_0001', '85078610_0008', '87702634_0007', '84686667_0002', '84802706_0002', '81811182_0004', '84421559_0005', '88221516_0007', '80057524_0005', '84209926_0018', '83959636_0010', '89989722_0016', '89225487_0007', '84391267_0001', '80889556_0002', '85250558_0011', '84567505_0005', '85814172_0007', '88884866_0005', '80497954_0012', '80666640_0014', '84939605_0004', '82141753_0018', '86874920_0014', '84505262_0010', '86288257_0001', '89699401_0001', '88537698_0013', '83958172_0001']

The corresponding director